### PR TITLE
Revert "Remove all"

### PR DESCRIPTION
**Pull request by mromanenkov**
This reverts commit 5d773626841dd73291f9870b8f35e344a42fa95d.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "compilerOptions": {
+    "outDir": "./public/",
```
Comment (vladimir-tikhonov) [inline tsconfig.json]: `strict: true`

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import Vector from './vector';
+
+export default class Utils {
+  public static isPointInPoly(point: number[], polyPoints: number[][]): boolean {
```
Comment (vladimir-tikhonov) [inline src/utils.ts]: please use `Vector` instead of array here and in other places

File: src/vector.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export default class Vector {
+  public x: number;
+  public y: number;
+
+  constructor(point: number[]) {
```
Comment (vladimir-tikhonov) [inline src/vector.ts]: Please change it to `x: number, y: number`

File: src/canvas.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import Polygon from './polygon';
+
+import ISetting from './interfaces/canvas-setting.interface';
+
+export default class Canvas {
+  id: string;
+  setting: ISetting;
+  objects: Polygon[];
+  selectedObject: Polygon;
+  element: any;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: Should be `HTMLElement`

File: src/canvas.ts [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    this.id = id;
+    this.setting = setting;
+    this.objects = objects || [];
+    this.selectedObject = null;
+    this.element = document.getElementById(id);
+    this.ctx = this.element.getContext('2d');
+    this.nextObjListPos = [this.setting.padding, this.setting.padding];
+  }
+
+  public init(): any {
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: it `void`. But you don't need to specify type here anyway

File: src/canvas.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  public add(object: Polygon): void {
+    this.objects.push(object);
+    object.setBoundingBox();
+
+    const offset = [this.nextObjListPos[0] - object.boundingBox[0][0],
+      this.nextObjListPos[1] - object.boundingBox[0][1]];
+
+    object.shift(offset);
+    object.setBoundingBox();
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: Why 2 updates of bounding box here? Anyway, it should be managed by `Polygon` itself

File: src/canvas.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    objectArr.forEach((object) => {
+      this.add(object);
+    });
+  }
+
+  public draw(object: Polygon, isFill?: boolean): void {
+    this.ctx.fillStyle = object.fillColor;
+    this.ctx.strokeStyle = object.strokeColor;
+
+    this.ctx.save();
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: `save` should be above `fillStyle` / `strokeStyle` changes, otherwise they will remain changed

File: src/canvas.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  public update(): void {
+    this.ctx.clearRect(0, 0, this.setting.width, this.setting.height);
+
+    this.findAllIntersectObjects();
+    this.checkVertexInPolyAll();
+
+    this.objects.forEach((object) => {
+      object.isOverlap ? this.draw(object, true) : this.draw(object);
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: I think that `draw` could check for `isOverlap` flag itself

File: src/canvas.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    });
+  }
+
+  public getSelectedObject(cursorPos: number[]): Polygon {
+    let selectedObject: Polygon;
+    this.objects.forEach((object) => {
+      const isInside = utils.isPointInPoly(cursorPos, object.points);
+      if (isInside) selectedObject = object;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: Not sure why linter hasn't catched this, but according to standard guidelines `if`'s body must be wrapped in `{..}`

File: src/canvas.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+
+  private checkVertexInPoly(polyA: Polygon, polyB: Polygon): boolean {
+    let isInPoly = false;
+
+    polyA.points.forEach((point) => {
+      isInPoly = utils.isPointInPoly(point, polyB.points);
+      if (isInPoly) {
+        isInPoly = true;
+        return false;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: `return` has no effect within `forEach`

File: src/canvas.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+      }
+    });
+    polyB.points.forEach((point) => {
+      isInPoly = utils.isPointInPoly(point, polyA.points);
+      if (isInPoly) {
+        isInPoly = true;
+        return false;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: ditto

File: src/canvas.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        if (isInPoly) {
+          objectA.isOverlap = true;
+          objectB.isOverlap = true;
+        }
+      }
+    }
+  }
+
+  private checkSideIntersection(polyA: Polygon, polyB: Polygon): boolean {
+    const pointsAcopy: number[][] = JSON.parse(JSON.stringify(polyA.points));
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: `JSON.stringify` is a really bad idea, please introduce `Polygon#clone` instead

File: src/canvas.ts [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+        isIntersect = utils.getIntersection(sideA, sideB);
+        if (isIntersect) return true;
+      }
+    }
+    return isIntersect;
+  }
+
+  private findOverlappedObject(polyA: Polygon, polyB: Polygon): boolean {
+    let isOverlap: boolean = false;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: the entire method could be replaced with `return this.checkSideIntersection(polyA, polyB) || this.checkVertexInPoly(polyA, polyB)`

File: src/canvas.ts [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    if (this.checkSideIntersection(polyA, polyB) || this.checkVertexInPoly(polyA, polyB)) {
+      isOverlap = true;
+    }
+    return isOverlap;
+  }
+
+
+  private findAllIntersectObjects(): void {
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: `Intersecting`

Comment (vladimir-tikhonov) [reply src/canvas.ts]: Moreover, the name is quite wrong, it doesn't really find anything, just updating polygon states

File: src/canvas.ts [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+    const overlapObjectIndeces: number[] = [];
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: It'll be better to use `Set` instead, you can add polygons directly into it

File: src/cursor.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+export default class Cursor {
+
+  private static instance: Cursor;
+
+  public cursorDownPos: number[] = [0, 0];
+  public cursorUpPos: number[] = [0, 0];
+
+  public static get Instance() {
+    return this.instance || (this.instance = new this());
```
Comment (vladimir-tikhonov) [inline src/cursor.ts]: `new this` looks quite bad, let it be `new Cursor()`

File: src/canvas.ts [hunk 17]
```diff
@@ -0,0 +1,177 @@
+import utils from './utils';
+import Polygon from './polygon';
+
+import ISetting from './interfaces/canvas-setting.interface';
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: Separate file for interface doesn't make much sense, you can keep it here instead

File: src/canvas.ts [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+        const objectB: Polygon = this.objects[j];
+
+        if (this.checkSideIntersection(objectA, objectB)) {
+          overlapingObjects.add(objectA).add(objectB);
+        }
+      }
+    }
+
+    this.objects.forEach((object) => {
+      object.isOverlap = overlapingObjects.has(object) ? true : false;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: ` ? true : false` is completely useless

File: src/cursor.ts [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+  public cursorDownPos = new Vector(0, 0);
+  public cursorUpPos = new Vector(0, 0);
+
+  public static get Instance() {
+    return this.instance || (this.instance = new Cursor());
+  }
+
+  public getOffset(): Vector {
+    const offset = new Vector(this.cursorUpPos.x - this.cursorDownPos.x,
```
Comment (vladimir-tikhonov) [inline src/cursor.ts]: `Vector#subtract`

File: src/canvas.ts [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+    this.checkVertexInPolyAll();
+
+    this.objects.forEach((object) => {
+      this.draw(object);
+    });
+  }
+
+  public getSelectedObject(cursorPos: Vector): Polygon | null {
+    let selectedObject: Polygon | null = null;
+    this.objects.forEach((object) => {
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: `find`

File: src/canvas.ts [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+      if (isInside) {
+        selectedObject = object;
+      }
+    });
+    return selectedObject;
+  }
+
+  private checkVertexInPoly(polyA: Polygon, polyB: Polygon): boolean {
+    let isInPoly = false;
+    polyA.vertices.forEach((vertex) => {
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: ditto

File: src/canvas.ts [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+          objectB.isOverlap = true;
+        }
+      }
+    }
+  }
+
+  private checkSideIntersection(polyA: Polygon, polyB: Polygon): boolean {
+    const polyAcopy: Polygon = polyA.clone();
+    const polyBcopy: Polygon = polyB.clone();
+    polyAcopy.vertices.push(polyAcopy.vertices[0]);
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: `[...polyA.vertices, polyA.vertices[0]]`

File: src/polygon.ts [hunk 23]
```diff
@@ -0,0 +1,45 @@
+import Vector from './vector';
+
+export default class Polygon {
+  public vertices: Vector[];
+  public strokeColor: string;
```
Comment (vladimir-tikhonov) [inline src/polygon.ts]: please move to `Canvas`

File: src/polygon.ts [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    this.vertices = vertices;
+    this.strokeColor = strokeColor || '#000';
+    this.fillColor = fillColor || '#fff';
+    this.isOverlap = false;
+    this.setBoundingBox();
+  }
+  
+  public shift(offset: Vector): void {
+    this.vertices.forEach((vertex) => {
+      vertex.x += offset.x;
```
Comment (vladimir-tikhonov) [inline src/polygon.ts]: `add`

File: src/polygon.ts [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+      vertex.y += offset.y;
+    });
+    this.setBoundingBox();
+  }
+  
+  public setBoundingBox(): void {
+    const minX: number = this.vertices.reduce((min, vertex) =>
```
Comment (vladimir-tikhonov) [inline src/polygon.ts]: please simply

File: src/polygon.ts [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+    const minY: number = this.vertices.reduce((min, vertex) =>
+        (vertex.y < min ? vertex.y : min),    this.vertices[0].y);
+    const maxY: number = this.vertices.reduce((min, vertex) =>
+        (vertex.y > min ? vertex.y : min),    this.vertices[0].y);
+        
+    this.boundingBox = [new Vector(minX, minY), new Vector(maxX, minY),
+      new Vector(maxX, maxY), new Vector(minX, maxY)];
+  }
+
+  public clone(): Polygon {
```
Comment (vladimir-tikhonov) [inline src/polygon.ts]: please remove

File: src/canvas.ts [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+  polygonMargin: number;
+}
+
+export class Canvas {
+  id: string;
+  setting: ISetting;
+  objects: Polygon[];
+  selectedObject: Polygon | null;
+  element: HTMLCanvasElement;
+  ctx: CanvasRenderingContext2D;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts]: please remove

File: tslint.json [hunk 28]
```diff
@@ -0,0 +1,9 @@
+{
+  "defaultSeverity": "error",
+  "extends": "tslint-config-airbnb",
+  "jsRules": {},
+  "rules": {
+    "import-name": false,
+    "no-increment-decrement": false
```
Comment (vladimir-tikhonov) [inline tslint.json]: please check

File: tslint.json [hunk 29]
```diff
@@ -0,0 +1,9 @@
+{
+  "defaultSeverity": "error",
+  "extends": "tslint-config-airbnb",
+  "jsRules": {},
+  "rules": {
+    "import-name": false,
```
Comment (vladimir-tikhonov) [inline tslint.json:6]: please check

Comment (mromanenkov) [reply tslint.json:6]: import-name rule requires matching of variable and file name

File: src/canvas.ts [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+  public addArr(objectArr: Polygon[]): void {
+    objectArr.forEach((object) => {
+      this.add(object);
+    });
+  }
+
+  public draw(object: Polygon): void {
+    const ctx = this.element.getContext('2d')!;
+    ctx.save();
+    ctx.fillStyle = object.isOverlap ? '#f00' : this.fillColor;
```
Comment (vladimir-tikhonov) [inline src/canvas.ts:62]: `tinycolor2('red')`

**Pull request merged**